table/block_based_table_reader.h

```diff
@@ -717,12 +733,17 @@ class BlockBasedTableIterator : public InternalIteratorBase<TValue> {
   bool for_compaction_;
   BlockHandle prev_index_value_;
 
-  static const size_t kInitReadaheadSize = 8 * 1024;
+  // All the below fields control iterator readahead
+  static const size_t kInitAutoReadaheadSize = 8 * 1024;
   // Found that 256 KB readahead size provides the best performance, based on
-  // experiments.
-  static const size_t kMaxReadaheadSize;
-  size_t readahead_size_ = kInitReadaheadSize;
+  // experiments, for auto readahead. Experiment data is in PR #3282.
+  static const size_t kMaxAutoReadaheadSize = 256 * 1024;
+  // If read_options.readahead_size is not set,
+  static const int kMinNumFileReadsToStartAutoReadahead = 2;
+  size_t readahead_size_ = 0;
+  size_t max_readahead_size_ = 0;
```

Comment:
Maybe I'm picky, but memory of BlockBasedTableIterator does matter a little bit. Not necessarily a big deal, but it's good to keep it small in long term. Larger memory consumption here may cause more memory allocation and worse cache locality. We should really think about keep it extra 8 bytes, rather than 20+. Can we just just do the if happening in iterator constructor in InitDataBlock()?

Reply:
Instead of fitting into the current model of exponential increasing buffer, by setting min and max to be the same, did you consider to just do it in a separate branch?
```
if (read_options_.readahead_size !=0 && num_file_reads_  == 0) {  // We already doing this branch anyway
      prefetch_buffer_.reset(new FilePrefetchBuffer(
              rep->file.get(), read_options_.readahead_size, read_options_.readahead_size));
} else {
......
```
We didn't do page cache readahead for `read_options.readahead_size > 0` for now so there is no reason to add more dependency on page cache.

Reply:
Here are the results of the benchmarks I ran: https://gist.github.com/sagar0/53693edc320a18abeaeca94ca32f5737 . 
For Buffered IO: If you look at the benchmarks which depend on Page Cache Prefetch vs. the ones that depend on our FilePrefetchBuffer, Page Cache Prefetch is more optimal than our buffer, as expected.

Reply:
Still there is no reason to add back a dependency on page cache.

Reply:
Also, users may already depend on the current behavior of user level reads. I would avoid a surprise to them. I have one use case in my mind, who may not be happy with starting to depend on readahead() with read_options.readahead_size. I can share with you offline.